transport/UdpEndpointImpl.cpp

```diff
@@ -252,32 +239,61 @@ void UdpEndpointImpl::registerListener(const std::string& stunUserName, IEvents*
 /** If using ICE, must be called from receive job queue to sync unregister */
 void UdpEndpointImpl::registerListener(const SocketAddress& srcAddress, IEvents* listener)
 {
-    auto dtlsIt = _dtlsListeners.find(srcAddress);
-    if (dtlsIt != _dtlsListeners.end())
+    auto it = _dtlsListeners.emplace(srcAddress, listener);
+    if (it.second)
+    {
+        logger::debug("register listener for %s", _name.c_str(), srcAddress.toString().c_str());
+        listener->onRegistered(*this);
+    }
+    else if (it.first != _dtlsListeners.cend() && it.first->second == listener)
+    {
+        // already registered
+    }
+    else
+    {
+        _receiveJobs.post(utils::bind(&UdpEndpointImpl::swapListener, this, srcAddress, listener));
+    }
+}
+
+void UdpEndpointImpl::swapListener(const SocketAddress& srcAddress, IEvents* newListener)
+{
+    auto it = _dtlsListeners.find(srcAddress);
+    if (it != _dtlsListeners.cend())
     {
         // src port is re-used. Unregister will look at listener pointer
-        if (dtlsIt->second == listener)
+        if (it->second == newListener)
         {
             return;
         }
 
-        if (dtlsIt->second)
+        if (it->second)
         {
-            _receiveJobs.addJob<UnRegisterNotifyListenerJob>(*this, *dtlsIt->second);
+            it->second->onUnregistered(*this);
         }
-        dtlsIt->second = listener;
-        listener->onRegistered(*this);
-    }
-    else
-    {
-        _dtlsListeners.emplace(srcAddress, listener);
-        listener->onRegistered(*this);
+        it->second = newListener;
+        newListener->onRegistered(*this);
+        return;
     }
+
+    logger::warn("dtls listener swap on %s skipped. Already removed", _name.c_str(), srcAddress.toString().c_str());
 }
 
 void UdpEndpointImpl::focusListener(const SocketAddress& remotePort, IEvents* listener)
 {
-    return;
+    _receiveJobs.post([this, remotePort, listener]() {
```

Comment:
I do not understand how skipping this code was helping. focusListener was not called in the failing test dtlsRace.